server/athenian/api/experiments/preloading/cache.py

```diff
@@ -0,0 +1,259 @@
+            self._mem["percentage"] = {
+                "series": self._mem["processed"]["series"]
+                / self._mem["raw"]["series"]
+                * 100,
+                "total": self._mem["processed"]["total"]
+                / self._mem["raw"]["total"]
+                * 100,
+            }
+
+        return self
+
+    async def refresh(self) -> "CachedDataFrame":
+        """Refresh the DataFrame from the database."""
+        if self._df is None:
+            raise ValueError("DataFrame not loaded")
+
+        return await self.load()
+
+    def memory_usage(
+        self, total: bool = False, human: bool = False,
+    ) -> Union[pd.Series, int, str]:
+        """Return information about the memory usage."""
+        s = self._df.memory_usage(index=True, deep=True)
+        if total:
+            size = sum(s)
+            return size if not human else _humanize_size(size)
+        else:
+            return s if not human else s.apply(_humanize_size)
+
+    def _squeeze(self, df: pd.DataFrame) -> pd.DataFrame:
+        def _cast_category(df, cols):
+            for c in cols:
+                df[c.name] = df[c.name].astype("category")
+            return df
+
+        def _cast_timestamp(df, cols):
+            for c in cols:
+                df[c.name] = (df[c.name].astype("int64") / 10 ** 9).astype("int32")
+                df.loc[df[c.name] < 0, c.name] = 0
+            return df
+
+        def _cast_identifiers(df, cols):
+            for c in cols:
+                max_length = df[c.name].apply(len).max()
+                df[c.name] = df[c.name].astype(f"|S{max_length}")
+            return df
+
+        return (
+            df.pipe(_cast_category, self._categorical_cols)
+            .pipe(_cast_timestamp, self._timestamp_cols)
+            .pipe(_cast_identifiers, self._identifier_cols)
+        )
+
+    def __repr__(self):
+        """Return the representation of the `CachedDataFrame`."""
+
+        def fmt_cols(cols):
+            return ", ".join(c.name for c in cols)
+
+        metadata = [
+            f"Table: {self._db.url}",
+            f"Columns: {fmt_cols(self._cols)}",
+            f"Timestamp columns: {fmt_cols(self._timestamp_cols)}",
+            f"Categorical columns: {fmt_cols(self._categorical_cols)}",
+            f"Identifier columns: {fmt_cols(self._identifier_cols)}",
+        ]
+
+        lines = metadata + ["DataFrame", self._df.__repr__()]
+        return "\n".join(lines)
+
+    def __getattr__(self, attr):
+        """Make attributes access propagate to the wrapped DataFrame."""
+        if self._df is None:
+            raise ValueError("DataFrame not loaded")
+
+        return self._df.__getattr__(attr)
+
+    def __getitem__(self, key):
+        """Make items access propagate to the wrapped DataFrame."""
+        if self._df is None:
+            raise ValueError("DataFrame not loaded")
+
+        return self._df.__getitem__(key)
```

Comment:
```suggestion
@property
def df(self) -> pd.DataFrame:
    assert self._df is not None, "Need to await refresh() at least once."
    return self._df
```
This way we don't need to proxy all dataframe's methods here.

Reply:
You're right, I was just too lazy to even type `.df` when playing in ipython 😅